### PR TITLE
Add random walks over UASTs

### PR DESCRIPTION
**Pull request by EgorBu**
* sourced/ml/uast_struct_to_bag.py - random walk logic
* sourced/ml/repo2/wmhash.py - extractor

### CODE REVIEW

File: sourced/ml/repo2/wmhash.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 from sourced.ml.repo2.base import Transformer
 from sourced.ml.uast_ids_to_bag import UastIds2Bag
+from sourced.ml.uast_struct_to_bag import UAST2RandomWalk2Bag
```
Comment (vmarkovtsev) [inline sourced/ml/repo2/wmhash.py]: Please rename to `UastRandomWalks2Bag`

File: sourced/ml/repo2/wmhash.py [hunk 2]
```diff
@@ -144,6 +145,36 @@ def inspect(self, uast):
             yield key
 
 
+class StructureBagExtractor(BagsExtractor):
```
Comment (vmarkovtsev) [inline sourced/ml/repo2/wmhash.py]: Really like this. Let's kill `extract` and `inspect` and merge `StructureBagExtractor` into `BagsExtractor`. Plus replace `self.uast2bag.uast_to_bag` with

```
def uast_to_bag(self, uast):
    raise NotImplementedError()
```

File: sourced/ml/uast_struct_to_bag.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+
+class UASTStructure2BagBase:
+    def uast_to_bag(self, uast):
+        raise NotImplemented
+
+
+class UAST2Seq2Bag(UASTStructure2BagBase):
+    """
+    DFS traversal + preserve order of node children.
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: Why DFS? As I remember we randomly walk.

File: sourced/ml/uast_struct_to_bag.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    """
+    def __init__(self, type2ind=None, stride=1, seq_len=5):
+        self.type2ind = type2ind if type2ind is not None else FakeVocabulary()
+        self.stride = stride
+        self.seq_len = seq_len
+
+    def _uast2seq(self, root, walk):
+        walk.append(root)
+        for ch in root.children:
+            self._uast2seq(ch, walk)
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: Please replace this with a non-recursive implementation

File: sourced/ml/uast_struct_to_bag.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        bag = defaultdict(int)
+        seq = []
+        self._uast2seq(uast, seq)
+
+        seq = [self.node2ind(n) for n in seq]
+
+        for i in range(0, len(seq) - self.seq_len, self.stride):
+            # convert to str - requirement from wmhash.BagsExtractor
+            bag[str(seq[i:i + self.seq_len])] += 1
+        return bag
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: Please place all the keys into `w.` namespace (see `IdentifiersBagExtractor`)

File: sourced/ml/uast_struct_to_bag.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        self.internal_type = internal_type
+        self.children = []
+
+
+class UAST2RandomWalks:
+    """
+    Generation of random walks for UAST.
+    """
+
+    def __init__(self, p, q, n_walks, n_steps, type2ind=None):
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: Please rename p and q into smth more verbose

File: sourced/ml/uast_struct_to_bag.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        options.extend(v.children)
+        return choice(options)
+
+
+class UAST2RandomWalk2Bag(UASTStructure2BagBase):
+    def __init__(self, p=0.5, q=0.5, n_walks=5, n_steps=19, stride=1, seq_len=[5, 6]):
+        self.random_walker = UAST2RandomWalks(p=p, q=q, n_walks=n_walks, n_steps=n_steps)
+        self.stride = stride
+
+        if (not isinstance(seq_len, int) and not isinstance(seq_len, list) and
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: `isinstance(seq_len, (int, list, tuple))`

File: sourced/ml/uast_struct_to_bag.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            return t
+
+        options = []
+        if v.parent is not None:
+            options.append(v.parent)
+        options.extend(v.children)
+        return choice(options)
+
+
+class UAST2RandomWalk2Bag(UASTStructure2BagBase):
```
Comment (vmarkovtsev) [inline sourced/ml/uast_struct_to_bag.py]: `@register_extractor`
`NAME = "node2vec"`

File: sourced/ml/uast_struct_to_bag.py (hunk 3 above)
Comment (EgorBu) [reply sourced/ml/uast_struct_to_bag.py]: it's another approach that traverses tree to preserve the order of nodes as it should be in code.

File: sourced/ml/uast_struct_to_bag.py (hunk 5 above)
Comment (EgorBu) [reply sourced/ml/uast_struct_to_bag.py]: Right now it's a part of `BagsExtractor` functionality - so every extractor should be provided with `namespace` argument during initialization.

**Pull request merged**